File: code/08-graphs/c/disjoint_set_demo.c
#include "graph.h"
#include <stdio.h>

int main(void) {
    DisjointSet d = {0};
    dsu_init(&d, 4);
    const char *name = "ABCD";
    size_t edges[][2] = {{0,1},{2,3},{1,2},{0,3}};
    for (size_t i = 0; i < 4; i++) {
        bool accepted = dsu_union(&d, edges[i][0], edges[i][1]);
        printf("%c%c: %s\n", name[edges[i][0]], name[edges[i][1]], accepted ? "accept" : "reject");
    }
    printf("components: ");
    for (size_t root = 0; root < 4; root++) {
        bool any = false;
        for (size_t v = 0; v < 4; v++) if (dsu_find(&d, v) == root) {
            printf(any ? ",%c" : "{%c", name[v]);
            any = true;
        }
        if (any) printf("}");
    }
    printf("\n");
    dsu_destroy(&d);
    return 0;
}

File: code/08-graphs/java/DisjointSetDemo.java
// Driver for the "DSU Animation" toy example (A=0, B=1, C=2, D=3): matches
// the trace's exact edge order and accept/reject decisions.
public final class DisjointSetDemo {
    public static void main(String[] args) {
        DisjointSet d = new DisjointSet(4);
        String[] name = {"A", "B", "C", "D"};
        int[][] edges = {{0,1},{2,3},{1,2},{0,3}};
        for (int[] e : edges) {
            boolean accepted = d.union(e[0], e[1]);
            System.out.println(name[e[0]] + name[e[1]] + ": " + (accepted ? "accept" : "reject"));
        }
        StringBuilder components = new StringBuilder();
        for (int root = 0; root < 4; root++) {
            boolean any = false;
            for (int v = 0; v < 4; v++) if (d.find(v) == root) {
                if (!any) { components.append("{"); any = true; } else components.append(",");
                components.append(name[v]);
            }
            if (any) components.append("}");
        }
        System.out.println("components: " + components);
    }
}
